README
INSTALLATION
============
Installation is pretty easy.

- ./configure
- Check Makefile.settings
- make
- make install


BACKGROUNDS
===========

The background of Rotix is a newsgroup where people liked to make jokes with
rot-xx obfuscations. At one point translating it by hand became to much of a
bother and I wrote rot, to do this for me.

rot was kind of a personal project and very primitive. I didn't bother much
for an interface, since I was the only person to use it. rot was a very
handy tool, but if I wanted to translate more than one line of text, it
would become annoying. So a new version was due.

At that time I was getting interested in Open-Source Software and in
releasing software. I wanted to do something back to the community who gave
me so much. Of course the project had to have a name: Rotix, derived from
the Dutch pronunciation of ROT-x. Rotix 0.7 was the result of this.


WEBSITE
=======

Of course Rotix has a website. You may even have downloaded this package
from it. The website contains an HTML-version of the Rotix man-page and some
of the cryptographic theories behind Rotix. The address is:

http://elektron.its.tudelft.nl/~hemmin98/rotix.html


LEGAL
=====

Rotix is distributed under the GPL (GNU General Public License). See the
file COPYING for this license.

	Rotix - A program to generate rotational obfuscations
	Copyright (C) 2001  Sjoerd Hemminga

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful, but
	WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
	General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
	USA


CHANGES
03 Apr 2002: Rotix 0.83

- Fixed a bug in the Makefile, which caused problems when compiling for
  FreeBSD
- Actually made -R/-L working. They were still bound to old test-switches
  (-d/-D).
- The ./configure now recognizes The Hurd and rotix has been reported to
  work.

06 Feb 2002: Rotix 0.82

- Fixed a bug in automated stdin-obfuscation.

27 Jan 2002: Rotix 0.81

- Added an option -a as alias for -r auto.
- Added a ./configure Thanks to Wilmer van der Gaast.
- Added some patches to increase portability.
- Added the possibility to automate obfuscation of stdin.

10 Nov 2001: Rotix 0.8

- Added command-line options that allows the user to control the direction
  of rotating (-R and -L).
- Added I18N support.
- Changed the --help output, so it fits on one screen.
- Fixed a bug, which allowed both stdin and a text to be specified on the
  command-line, when entered in a certain sequence.

13 Oct 2001: Rotix 0.71
- Fixed a bug, which causes blank lines to have the contents of the previous
  line.

18 Sep 2001: Rotix 0.7
- First release.


rot.c
/*\
 *	Rotix - A program to generate rotational obfuscations
 *	Copyright (C) 2001  Sjoerd Hemminga
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; either version 2 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful, but
 *	WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *	General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 *	USA
\*/


/*\
 *  Define the flags used.
 *
 *  This is upwards compatible. Programs using this interface will still be
 *  able to connect to later versions of this file. Other flags will be
 *  added when needed.
\*/

#define ROTATE_RIGHT_FLAG	1

/*\
 *  This function rotates rotar by rotor. NOTE that rotar contains the
 *  rotated contents after running this function.
\*/

void rotate (int rotor, char *rotar, char flags)
{
	int i;
	
	if (!(flags & ROTATE_RIGHT_FLAG)) {
		/* Instead of rotate right, rotate left. */
		rotor = -rotor;
	}
	
	/* Make sure that rotor-values over 127 don't cause problems. */
	rotor %= 26;
	
	/* Convert negative rotor values to a positive equivalent. */
	if (rotor < 0) {
		rotor += 26;
	}
	
	for (i = 0; rotar[i]; i++) {
		if ( (rotar[i] >= 65) && (rotar[i] <= 90) ) {
			rotar[i] -= 65;
			rotar[i] += rotor;
			rotar[i] %= 26;
			rotar[i] += 65;
		}
		
		if ( (rotar[i] >= 97) && (rotar[i] <= 122) ) {
			rotar[i] -= 97;
			rotar[i] += rotor;
			rotar[i] %= 26;
			rotar[i] += 97;
		}
	}
}


help.c
/*\
 *	Rotix - A program to generate rotational obfuscations
 *	Copyright (C) 2001  Sjoerd Hemminga
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; either version 2 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful, but
 *	WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *	General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 *	USA
\*/

#include <stdio.h>

/*\
 * I18N
\*/

#include <libintl.h>
#define _(String) gettext (String)

/*\
 *  This function displays the help-information.
\*/

void display_help ()
{
	printf (_("Usage:\n"));
	printf (_("       rotix [ -f <file> ] [ -o <file> ] [ -r <value> ]\n"));
	printf (_("       rotix -t <text> [ -o <file> ] [ -r <value> ]\n\n"));

	printf (_("-f <file>  specify input-file\n"));
	printf (_("-o <file>  specify output-file\n"));
	printf (_("-r <value> specify the value to rotate with\n"));
	printf (_("-a         use all sane values to rotate with\n"));
	printf (_("-R         rotate right\n"));
	printf (_("-L         rotate left\n"));
	printf (_("-t <text>  specify the text to rotate\n\n"));
	
	printf (_("For complete documentation:\n"));
	printf ("man rotix\n\n");

	printf (_("Report bugs to: sjoerd@huiswerkservice.nl\n"));
}

void display_version ()
{
	printf (_("Rotix version 0.82\n\n"));

	printf ("Copyright (C) 2001 Sjoerd Hemminga <sjoerd@huiswerkservice.nl>\n\n");

	printf (_("Rotix is licensed under the GNU General Public License. See the file COPYING\n"
		  "for details.\n"));
}


rotix.c
/*\
 *	Rotix - A program to generate rotational obfuscations
 *	Copyright (C) 2001  Sjoerd Hemminga <sjoerd@huiswerkservice.nl>
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; either version 2 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful, but
 *	WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *	General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 *	USA
\*/


/* Rotix includes. */
#include "rot.h"
#include "help.h"

/* System includes. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef NO_GETOPT_LONG
#include <unistd.h>
#define getopt_long(a, b, c, d, e) getopt(a, b, c)
#else
#include <getopt.h>
#endif

/* I18N */
#ifdef I18N
#include <libintl.h>
#define _(String) gettext (String)
#else
#define _(String) String
#endif

/* Defining the flags. */
#define ROT_AUTO		1
#define ROT_INPUT_FILE		2
#define ROT_TEXT		4
#define ROT_OUTPUT_FILE		8
#define ROT_STDIN		16

int main (int argc, char **argv)
{
	int i = 1, rotor = 13, option = 0;
	char *arg = NULL, option_flags = 0, rotate_flags = 0;
	
#ifndef NO_GETOPT_LONG
	/* The struct used to determine which options are called. */
	static struct option rotix_options[] = {
		{ "file",	1,	NULL,	'f' },
		{ "rot",	1,	NULL,	'r' },
		{ "auto",	0,	NULL,	'a' },
		{ "output",	1,	NULL,	'o' },
		{ "text",	1,	NULL,	't' },
		{ "left",	0,	NULL,	'L' },
		{ "right",	0,	NULL,	'R' },
		{ "help",	0,	NULL,	'h' },
		{ "version",	0,	NULL,	'v' },
		{ NULL,		0,	NULL,	 0  }
	};
#endif


	/* A 2-bit character array, to obfuscate a character */
	char rotme[2];
	
	FILE *ifp = NULL;
	FILE *ofp = NULL;

#ifdef I18N
	/* Initialize I18n */
	setlocale (LC_MESSAGES, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);
#endif

	/* Initialization of some variables */	
	rotme[1] = '\0';
	rotate_flags |= ROTATE_RIGHT_FLAG;

/* Parsing options. */
	while (1) {
		option = getopt_long (argc, argv, "f:r:ao:t:LRhv", rotix_options, NULL);
		if (option == -1) break;
		
		switch (option) {
			case '?':
				display_help ();
				exit (1);
			case ':':

			/* Missing argument. */
				display_help ();
				exit (1);
			case 'v':
				display_version ();
				exit (0);
			case 'h':
				display_help ();
				exit (0);

			/* Options that affect the running of Rotix. */
			
			/* Direction right */
			case 'R':
				rotate_flags |= ROTATE_RIGHT_FLAG;
			break;
			
			case 'L':
				if (rotate_flags & ROTATE_RIGHT_FLAG) {
					rotate_flags -= 1;
				}
			break;

			/* Input file */
			case 'f':
				if (strcmp (optarg, "-") == 0) {
					ifp = stdin;
					option_flags |= ROT_STDIN;
					
					if (option_flags & ROT_TEXT) {
						fprintf (stderr, _("rotix: you cannot specify both an inputfile and a text\n"));
						exit (1);
					}
				} else {
					ifp = fopen (optarg, "r");
					option_flags |= ROT_INPUT_FILE;

					if (ifp == NULL) {
						fprintf (stderr, _("rotix: error opening file %s for reading\n"), optarg);
						exit (1);
					} else if (option_flags & ROT_TEXT) {
						fprintf (stderr, _("rotix: you cannot specify both an inputfile and a text\n"));
						exit (1);
					}
				}
			break;

			/* Rotate number */
			case 'r':
				if (strcmp (optarg, "auto") == 0) {
					option_flags |= ROT_AUTO;
				} else if ( !sscanf (optarg, "%9i", &rotor) ) {
					display_help ();
					exit (1);
				}
			break;
			
			/* Automated rotating */
			case 'a':
				option_flags |= ROT_AUTO;
			break;

			/* Output file */
			case 'o':
				if (strcmp (optarg, "-") == 0) {
					ofp = stdout;
				} else {
					ofp = fopen (optarg, "a");

					option_flags |= ROT_OUTPUT_FILE;
					
					if (ofp == NULL) {
						fprintf (stderr, _("rotix: error opening file %s for writing\n"), optarg);
						exit (1);
					}
				}
			break;

			/* Text */
			case 't':
				if ((option_flags & ROT_INPUT_FILE) || (option_flags & ROT_STDIN)) {
					fprintf (stderr, _("rotix: you cannot specify both an inputfile and a text\n"));
					exit (1);
				}
				
				arg = (char *)malloc (strlen (optarg) + 1);
				if (arg == NULL) {
					fprintf (stderr, _("rotix: not enough memory\n"));
					exit (1);
				}
				strcpy (arg, optarg);
				option_flags |= ROT_TEXT;
			break;
			
			/* Unrecognized option. NOTE: this should never happen, since case '?' should catch this one */
			default:
				fprintf (stderr, _("rotix: fixme\n"));
				exit (2);
		}
	}

	if (!(option_flags & ROT_INPUT_FILE) && !(option_flags & ROT_TEXT)) {
		option_flags |= ROT_STDIN;
		ifp = stdin;
	}
	
	if (ofp == NULL) {
		ofp = stdout;
	}
	
/*\
 *  End of option-parsing.
\*/
	if (option_flags & ROT_TEXT) {
		if (option_flags & ROT_AUTO) {
			for (i = 1; i < 26; i++) {
				rotate (1, arg, rotate_flags);
				fprintf (ofp, "ROT-%02i: %s\n", i, arg);
			}
			
			free (arg);
		} else {
			rotate (rotor, arg, rotate_flags);
			fprintf (ofp, "%s\n", arg);
		}
	} else if (option_flags & ROT_AUTO) {
		/* Goodie, goodie, automated (de-)obfuscation of a file! */
		
		if (ifp == stdin) {
			/* Very dirty hack to get this functioning without using any extra variables */
			ifp = tmpfile ();
			
			if (ifp == NULL) {
				fprintf (stderr, _("rotix: error creating temporary file\n"));
				exit (1);
			}
			
			arg = (char *)malloc (265);
			if (arg == NULL) {
				fprintf (stderr, _("rotix: not enough memory\n"));
				exit (1);
			}
			
			while ( !feof (stdin) ) {
				strcpy (arg, "");
				rotme[0] = '\0';
				
				fscanf (stdin, "%255[^\n]s", arg);
				fprintf (ifp, "%s", arg);
				
				if (feof (stdin)) break;
				
				rotme[0] = fgetc (stdin);
				fprintf (ifp, "%s", rotme);
			}
			rewind (ifp);
		}
		
		arg = (char *)malloc (256);
		if (arg == NULL) {
			fprintf (stderr, _("rotix: not enough memory\n"));
			exit (1);
		}

		for (i = 1; i < 26; i++) {
			fprintf (ofp, "\nROT-%02i:\n=======\n\n", i);
			
			while ( !feof (ifp) ) {
				strcpy (arg, "");
				rotme[0] = '\0';

				fscanf (ifp, "%255[^\n]s", arg);
				rotate (i, arg, rotate_flags);
				fprintf (ofp, "%s", arg);

				if (feof (ifp)) break;

				rotme[0] = fgetc (ifp);
				rotate (i, rotme, rotate_flags);
				fprintf (ofp, "%s", rotme);
			}
			rewind (ifp);
		}
		free (arg);
	} else {
		arg = (char *)malloc (256);
		if (arg == NULL) {
			fprintf (stderr, _("rotix: not enough memory\n"));
			exit (1);
		}

		while ( !feof (ifp) ) {
			strcpy (arg, "");
			rotme[0] = '\0';

			fscanf (ifp, "%255[^\n]s", arg);
			rotate (rotor, arg, rotate_flags);
			fprintf (ofp, "%s", arg);

			if (feof (ifp)) break;

			rotme[0] = fgetc (ifp);
			rotate (rotor, rotme, rotate_flags);
			fprintf (ofp, "%s", rotme);
		}

		free (arg);
	}

	fflush (NULL);

	if (option_flags & ROT_INPUT_FILE) {
		fclose (ifp);
	}
	
	if (option_flags & ROT_OUTPUT_FILE) {
		fclose (ofp);
	}		
	
	return (0);
}
